drivers/vulkan/rendering_device_driver_vulkan.cpp

```diff
@@ -2684,43 +2770,13 @@ void RenderingDeviceDriverVulkan::_swap_chain_release(SwapChain *swap_chain) {
 RenderingDeviceDriver::SwapChainID RenderingDeviceDriverVulkan::swap_chain_create(RenderingContextDriver::SurfaceID p_surface) {
 	DEV_ASSERT(p_surface != 0);
 
-	RenderingContextDriverVulkan::Surface *surface = (RenderingContextDriverVulkan::Surface *)(p_surface);
-	const RenderingContextDriverVulkan::Functions &functions = context_driver->functions_get();
-
-	// Retrieve the formats supported by the surface.
-	uint32_t format_count = 0;
-	VkResult err = functions.GetPhysicalDeviceSurfaceFormatsKHR(physical_device, surface->vk_surface, &format_count, nullptr);
-	ERR_FAIL_COND_V(err != VK_SUCCESS, SwapChainID());
-
-	TightLocalVector<VkSurfaceFormatKHR> formats;
-	formats.resize(format_count);
-	err = functions.GetPhysicalDeviceSurfaceFormatsKHR(physical_device, surface->vk_surface, &format_count, formats.ptr());
-	ERR_FAIL_COND_V(err != VK_SUCCESS, SwapChainID());
-
+	// Determine the format and color space for the swap chain.
 	VkFormat format = VK_FORMAT_UNDEFINED;
 	VkColorSpaceKHR color_space = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR;
-	if (format_count == 1 && formats[0].format == VK_FORMAT_UNDEFINED) {
-		// If the format list includes just one entry of VK_FORMAT_UNDEFINED, the surface has no preferred format.
-		format = VK_FORMAT_B8G8R8A8_UNORM;
-		color_space = formats[0].colorSpace;
-	} else if (format_count > 0) {
-		// Use one of the supported formats, prefer B8G8R8A8_UNORM.
-		const VkFormat preferred_format = VK_FORMAT_B8G8R8A8_UNORM;
-		const VkFormat second_format = VK_FORMAT_R8G8B8A8_UNORM;
-		for (uint32_t i = 0; i < format_count; i++) {
-			if (formats[i].format == preferred_format || formats[i].format == second_format) {
-				format = formats[i].format;
-				if (formats[i].format == preferred_format) {
-					// This is the preferred format, stop searching.
-					break;
-				}
-			}
-		}
+	if (!_determine_swap_chain_format(p_surface, format, color_space)) {
+		ERR_FAIL_V_MSG(SwapChainID(), "Surface did not return any valid formats.");
 	}
 
-	// No formats are supported.
-	ERR_FAIL_COND_V_MSG(format == VK_FORMAT_UNDEFINED, SwapChainID(), "Surface did not return any valid formats.");
-
 	// Create the render pass for the chosen format.
 	VkAttachmentDescription2KHR attachment = {};
 	attachment.sType = VK_STRUCTURE_TYPE_ATTACHMENT_DESCRIPTION_2_KHR;
```

Comment:
`attachment.format = format` (one line below) also changes depending on the status of HDR. This is used to create the render pass (`swap_chain->render_pass` down below) yet I don't see anything that updates the render pass when HDR is toggled.

Honestly, I don't know how the render pass is used, so I don't know if it is necessary to update, what to update and how to update. Does anyone know?

Reply:
I'm not sure either, things seem to work fine without updating the render pass. The render pass is only created once when the swap chain is created, then never updated again, even as the swap chain's images are rebuilt on resize.

It would be nice to get some feedback from the rendering team on why this is done this way and if it would be fine to rebuild the rendering pass when HDR mode is changed.

Reply:
We always use dynamic viewport and scissors so render passes never need to be recreated if the dimension of targets change.

The specification describes that a framebuffer and a render pass is compatible if the format matches. If the format changes, then it is necessary to recreate the render pass.

https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap8.html#renderpass-compatibility

> Honestly, I don't know how the render pass is used

It's used for anything that starts rendering to the screen directly, although right now it is only used for the final blit of the framebuffer to the swap chain.

Reply:
This should be fixed now.

When the swap chain is invalidated, the render pass is now recreated to use the new format. I also needed to updated the compositor to rebuild its pipelines to account for the new format. After these changes, I no longer see any vulkan validation errors related to swapping HDR output mode.